### PR TITLE
Add basic Kokkos support for the original tests of #783

### PR DESCRIPTION
**Pull request by gojakuch**
This commit adds a test from the original Kokkos PR #783 and provides the implementation of the simplest Kokkos features in the form of custom pushforwards. This commit only addresses the forward mode.

### CODE REVIEW

**Comment by gojakuch** [comment]
The exact way this has been implemented in this PR is a subject for a discussion. I'm specifically talking about the way I put all the pushforwards in this file in the unittest directory. I also removed the initial tests I'd written in favour of the original ones from the PR #783 that basically cover the same things.

**Comment by gojakuch** [comment]
@vgvassilev as far as I can see, clang-format and some other checks are failing because there's no valid OpenPGP data being found, so it's probably not the PR's fault.

**Comment by MihailMihov** [comment]
@gojakuch It couldn't resolve the url (`curl: (6) Could not resolve host: apt.llvm.org`), but the url is working now, so maybe just re-run the checks?

**Comment by gojakuch** [comment]
@MihailMihov thanks, I'll do that a bit later.

File: unittests/Kokkos/KokkosImplementation.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
+// This header file contains the implementation of the Kokkos framework
```
Comment (vgvassilev) [inline unittests/Kokkos/KokkosImplementation.h:1]: This file should become `KokkosBuiltins.h`, similar to `[STLBuiltins.h](https://github.com/vgvassilev/clad/blob/master/include/clad/Differentiator/STLBuiltins.h)`. In it we should add the propagators which work with kokkos. This way users should be able to include this file and get them ready to go.

Comment (gojakuch) [reply unittests/Kokkos/KokkosImplementation.h:1]: yes, this is exactly the part of the design that I was unsure about. I'll move it as I fix the errors in the checks

File: unittests/Kokkos/CMakeLists.txt [hunk 2]
```diff
@@ -1,6 +1,6 @@
 add_clad_unittest(KokkosTests
   main.cpp
-  ViewBasics.cpp
```
Comment (vgvassilev) [inline unittests/Kokkos/CMakeLists.txt:3]: I'd keep the ViewBasics as it has a lot of commented out code that should soon work.

**Comment by gojakuch** [comment]
@vgvassilev the failing check I have right now in this PR is the most important Kokkos one and yet I cannot replicate it on my machine. it works out fine for me. what shall I do in this situation?

**Comment by vgvassilev** [comment]
> @vgvassilev the failing check I have right now in this PR is the most important Kokkos one and yet I cannot replicate it on my machine. it works out fine for me. what shall I do in this situation?

Debug on the bots?

**Comment by gojakuch** [comment]
> Debug on the bots?

how do I do that?

**Comment by gojakuch** [comment]
@vgvassilev so basically I've managed to debug and fix this on the GitHub runner, but only by modifying a line of code in Clad's main code base. the change is quite small and honestly seems reasonable to me, but that's still something I'd like to discuss during today's meeting. I'm pushing it here for now to have all the checks passing, but maybe it'd be better to have it as a separate commit.

File: include/clad/Differentiator/KokkosBuiltins.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// This header file contains the implementation of the Kokkos framework
+// differentiation support in Clad in the form of custom pushforwards and
+// pullbacks. Please include it manually to use Clad on Kokkos code.
+
+#ifndef CLAD_KOKKOS_BUILTINS_H
+#define CLAD_KOKKOS_BUILTINS_H
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/KokkosBuiltins.h:6]: That's a good suggestion.

File: include/clad/Differentiator/KokkosBuiltins.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+#include <Kokkos_Core.hpp>
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/KokkosBuiltins.h:8]: Maybe wrap this in `_has_include`?

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 5]
```diff
@@ -1198,7 +1198,7 @@ StmtDiff BaseForwardModeVisitor::VisitCallExpr(const CallExpr* CE) {
   // If all arguments are constant literals, then this does not contribute to
   // the gradient.
   // FIXME: revert this when this is integrated in the activity analysis pass.
-  if (!callDiff) {
+  if (!callDiff && !CE->getType().getTypePtr()->isVoidType()) {
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1201]: We should move this as a separate PR with an appropriate test.

**Review by vgvassilev — COMMENTED** [review]
I think this is heading in a good direction.

@kliegeois, @brian-kelley can you take a look?

File: include/clad/Differentiator/KokkosBuiltins.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+#include "clad/Differentiator/Differentiator.h"
+
+namespace clad {
+namespace custom_derivatives {
+namespace class_functions {
+/// Kokkos arrays
+template <typename ArrayT, typename Space>
+clad::ValueAndPushforward<Kokkos::View<ArrayT, Space>,
+                          Kokkos::View<ArrayT, Space>>
+constructor_pushforward(
```
Comment (brian-kelley) [inline include/clad/Differentiator/KokkosBuiltins.h:18]:
```suggestion
template <class DataType, class... ViewParams>
clad::ValueAndPushforward<Kokkos::View<DataType, ViewParams...>,
                          Kokkos::View<Kokkos::View<DataType, ViewParams...>>>
constructor_pushforward(
```

The full set of possible ``Kokkos::View`` template parameters is more than ``<datatype, space>`` or ``<datatype, layout, space>``. There could also be a memory traits type, and everything besides the datatype is optional. What about doing something like this to match all the possible cases? This would also let you merge this specialization with the one on line 32.

**Review by kliegeois — COMMENTED** [review]
Thanks @gojakuch it looks good to me!

**Comment by brian-kelley** [comment]
@gojakuch It also looks good to me, I just see that using variadic template arguments could simplify the ``constructor_pushforward`` specializations and support more general ``Kokkos::View`` types.

File: include/clad/Differentiator/KokkosBuiltins.h (hunk 6 above)
Comment (kliegeois) [reply include/clad/Differentiator/KokkosBuiltins.h:18]: Good catch @brian-kelley !

File: unittests/Kokkos/ViewAccess.cpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
+// Source:
+// https://github.com/vgvassilev/clad/pull/783/files#diff-c919937640674e1911d1f17d3df2b65426d152400589873ff9ba2f0391048d78
+// Source: https://github.com/kliegeois/clad/tree/kokkos-PR
```
Comment (vgvassilev) [inline unittests/Kokkos/ViewAccess.cpp:3]:
```suggestion
```

File: unittests/Kokkos/ViewBasics.cpp [hunk 8]
```diff
@@ -4,12 +4,12 @@
 // it has been modified to match gtest guidelines and improve readability
 
 #include "ParallelAdd.h"
-#include "TestUtils.h"
+// #include "TestUtils.h"
```
Comment (vgvassilev) [inline unittests/Kokkos/ViewBasics.cpp:7]: Why this is commented out? Maybe we should remove it?

File: unittests/Kokkos/ViewBasics.cpp [hunk 9]
```diff
@@ -4,12 +4,12 @@
...
 #include <Kokkos_Core.hpp>
 #include "clad/Differentiator/Differentiator.h"
 #include "gtest/gtest.h"
 
-double f(double x, double y) {
+double ff(double x, double y) {
```
Comment (vgvassilev) [inline unittests/Kokkos/ViewBasics.cpp:12]: I'd love to use more meaningful names.

Comment (gojakuch) [reply unittests/Kokkos/ViewBasics.cpp:12]: alright, I'll rename these. it's just that I couldn't use the same name twice, because this google test thing complains about having functions with the same name, as it compiles all the tests at once, not like clad does with the other tests.

File: include/clad/Differentiator/KokkosBuiltins.h (hunk 6 above)
Comment (gojakuch) [reply include/clad/Differentiator/KokkosBuiltins.h]: thanks for the suggestion, updated this.

**Comment by gojakuch** [comment]
@vgvassilev the `ubu22-clang16-runtime18` check started failing both here and on another PR of mine for some reason related with package setup there, so that's not on my side

File: include/clad/Differentiator/KokkosBuiltins.h [hunk 10]
```diff
@@ -0,0 +1,55 @@
+// This header file contains the implementation of the Kokkos framework
+// differentiation support in Clad in the form of custom pushforwards and
+// pullbacks. Please include it manually to use Clad on Kokkos code.
```
Comment (vgvassilev) [inline include/clad/Differentiator/KokkosBuiltins.h:3]:
```suggestion
// pullbacks. Please include it manually to enable Clad for Kokkos code.
```

File: include/clad/Differentiator/KokkosBuiltins.h [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+#ifndef CLAD_KOKKOS_BUILTINS_H
+#define CLAD_KOKKOS_BUILTINS_H
+
+#include <Kokkos_Core.hpp>
+#include "clad/Differentiator/Differentiator.h"
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/KokkosBuiltins.h:9]: We can ignore that suggestion.

File: include/clad/Differentiator/KokkosBuiltins.h [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+          Kokkos::View<ArrayT, Layout, Space>("_diff_" + name, idx0, idx1, idx2,
+                                              idx3, idx4, idx5, idx6, idx7)};
+}
+} // namespace class_functions
+
+/// Kokkos functions
+namespace Kokkos {
+template <typename View1, typename View2, typename T>
+inline void deep_copy_pushforward(const View1& dst, const View2& src, T param,
+                                  const View1& _d_dst, const View2& _d_src,
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/KokkosBuiltins.h:36]: These suggestions are relevant. The `_` identifiers are normally reserved for special builtins and we are not doing anything special here.

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Comment by kliegeois** [comment]
Thanks a lot for having merged the PR @vgvassilev and @gojakuch !
